File: desktop-prototypes/build-atom-python/atom_model.py
ELEMENTS = [
    ("?", "No element"),
    ("H", "Hydrogen"),
    ("He", "Helium"),
    ("Li", "Lithium"),
    ("Be", "Beryllium"),
    ("B", "Boron"),
    ("C", "Carbon"),
    ("N", "Nitrogen"),
    ("O", "Oxygen"),
    ("F", "Fluorine"),
    ("Ne", "Neon"),
]

MAX_PROTONS = 10
MAX_ELECTRONS = 10


class AtomModel:
    """Minimal model for the Build an Atom desktop prototype."""

    def __init__(self):
        self.reset()

    def add_proton(self):
        if self.protons < MAX_PROTONS:
            self.protons += 1

    def add_neutron(self):
        self.neutrons += 1

    def add_electron(self):
        if self.electrons < MAX_ELECTRONS:
            self.electrons += 1

    def reset(self):
        self.protons = 0
        self.neutrons = 0
        self.electrons = 0

    def atomic_number(self):
        return self.protons

    def mass_number(self):
        return self.protons + self.neutrons

    def charge(self):
        return self.protons - self.electrons

    def element_symbol(self):
        if 0 <= self.protons < len(ELEMENTS):
            return ELEMENTS[self.protons][0]
        return "?"

    def element_name(self):
        if 0 <= self.protons < len(ELEMENTS):
            return ELEMENTS[self.protons][1]
        return "Unknown"

    def electron_shell_counts(self):
        shell_one = min(self.electrons, 2)
        shell_two = max(0, min(self.electrons - shell_one, 8))
        return shell_one, shell_two

File: portable-core/build-atom/AtomModel.cpp
#include "AtomModel.h"

namespace {
constexpr const char* ElementSymbols[] = {
    "?",
    "H",
    "He",
    "Li",
    "Be",
    "B",
    "C",
    "N",
    "O",
    "F",
    "Ne",
};

constexpr const char* ElementNames[] = {
    "Unknown",
    "Hydrogen",
    "Helium",
    "Lithium",
    "Beryllium",
    "Boron",
    "Carbon",
    "Nitrogen",
    "Oxygen",
    "Fluorine",
    "Neon",
};
}

void AtomModel::reset() {
    protons = 0;
    neutrons = 0;
    electrons = 0;
}

bool AtomModel::addProton() {
    if (protons >= MaxProtons) {
        return false;
    }

    ++protons;
    return true;
}

bool AtomModel::addNeutron() {
    if (neutrons >= MaxNeutrons) {
        return false;
    }

    ++neutrons;
    return true;
}

bool AtomModel::addElectron() {
    if (electrons >= ShellOneCapacity + ShellTwoCapacity) {
        return false;
    }

    ++electrons;
    return true;
}

bool AtomModel::removeProton() {
    if (protons <= 0) {
        return false;
    }

    --protons;
    return true;
}

bool AtomModel::removeNeutron() {
    if (neutrons <= 0) {
        return false;
    }

    --neutrons;
    return true;
}

bool AtomModel::removeElectron() {
    if (electrons <= 0) {
        return false;
    }

    --electrons;
    return true;
}

int AtomModel::getProtons() const {
    return protons;
}

int AtomModel::getNeutrons() const {
    return neutrons;
}

int AtomModel::getElectrons() const {
    return electrons;
}

int AtomModel::getAtomicNumber() const {
    return protons;
}

int AtomModel::getMassNumber() const {
    return protons + neutrons;
}

int AtomModel::getCharge() const {
    return protons - electrons;
}

const char* AtomModel::getElementSymbol() const {
    return symbolForProtons(protons);
}

const char* AtomModel::getElementName() const {
    return nameForProtons(protons);
}

bool AtomModel::isNeutral() const {
    return getCharge() == 0;
}

const char* AtomModel::symbolForProtons(int count) {
    if (count < 0 || count > MaxProtons) {
        return "?";
    }

    return ElementSymbols[count];
}

const char* AtomModel::nameForProtons(int count) {
    if (count < 0 || count > MaxProtons) {
        return "Unknown";
    }

    return ElementNames[count];
}

